### DISCUSSION TITLE
Use Library Header in Fprime Submodule

### DISCUSSION THREAD

**Discussion by legounicycler** [nasa/fprime#1070] [Q&A]
Another question from the GT-2 team. We are running into the following issue:

The PolyTypes.cpp file uses macros with the form PRI___ (https://github.com/nasa/fprime/blob/devel/Fw/Types/PolyType.cpp#L617)

These are being defined by the standard library header inttypes.h included by polytypes here: https://github.com/nasa/fprime/blob/devel/Fw/Types/PolyType.cpp#L5

Arduino has an alternative definition for this file which uses double quotes instead of single quotes in its macro definitions: https://github.com/codebendercc/arduino-core-files/blob/master/v102/hardware/tools/avr/lib/avr/include/inttypes.h

The difference between single and double quotes is causing build errors.

The question is, how do we make sure components within the fprime submodule link against the Arduino header instead of the C standard library header?

Note: We are hoping to avoid modifying the submodule. We link our custom components to arduino by modifying each component's CMakeLists.txt file. 

CC: @tumbleshack @SterlingPeet @teddylambert

**Comment by timcanham** [comment]
I'm a little confused by this - I looked at the Linux `inttypes.h` macros, and they also have double quotes.

**Reply by Joshua-Anderson** [reply]
Agreed - double quotes are the expectation. What compilation error are you seeing?

**Reply by legounicycler** [reply]
The following error shows up when running fprime-util build:

/media/sf_SharedFolder/gt2-software/fprime/Fw/Types/PolyType.cpp: In member function 'void Fw::PolyType::toString(Fw::StringBase&, bool) const':
/media/sf_SharedFolder/gt2-software/fprime/Fw/Types/PolyType.cpp:617:67: error: expected ')' before 'PRIu8'
                 (void) snprintf(valString, sizeof(valString), "%" PRIu8 " ", this->m_val.u8Val);
                                                                   ^~~~~
/media/sf_SharedFolder/gt2-software/fprime/Fw/Types/PolyType.cpp:620:67: error: expected ')' before 'PRId8'
                 (void) snprintf(valString, sizeof(valString), "%" PRId8 " ", this->m_val.i8Val);
                                                                   ^~~~~
/media/sf_SharedFolder/gt2-software/fprime/Fw/Types/PolyType.cpp:624:67: error: expected ')' before 'PRIu16'
                 (void) snprintf(valString, sizeof(valString), "%" PRIu16 " ", this->m_val.u16Val);
                                                                   ^~~~~~
/media/sf_SharedFolder/gt2-software/fprime/Fw/Types/PolyType.cpp:627:67: error: expected ')' before 'PRId16'
                 (void) snprintf(valString, sizeof(valString), "%" PRId16 " ", this->m_val.i16Val);
                                                                   ^~~~~~
/media/sf_SharedFolder/gt2-software/fprime/Fw/Types/PolyType.cpp:632:67: error: expected ')' before 'PRIu32'
                 (void) snprintf(valString, sizeof(valString), "%" PRIu32 " ", this->m_val.u32Val);
                                                                   ^~~~~~
/media/sf_SharedFolder/gt2-software/fprime/Fw/Types/PolyType.cpp:635:67: error: expected ')' before 'PRId32'
                 (void) snprintf(valString, sizeof(valString), "%" PRId32 " ", this->m_val.i32Val);
                                                                   ^~~~~~
/media/sf_SharedFolder/gt2-software/fprime/Fw/Types/PolyType.cpp:640:67: error: expected ')' before 'PRIu64'
                 (void) snprintf(valString, sizeof(valString), "%" PRIu64 " ", this->m_val.u64Val);
                                                                   ^~~~~~
/media/sf_SharedFolder/gt2-software/fprime/Fw/Types/PolyType.cpp:643:64: error: expected ')' before 'PRId64'
F-Prime/Fw/Types/CMakeFiles/Fw_Types.dir/build.make:230: recipe for target 'F-Prime/Fw/Types/CMakeFiles/Fw_Types.dir/PolyType.cpp.obj' failed
              (void) snprintf(valString, sizeof(valString), "%" PRId64 " ", this->m_val.i64Val);
                                                                ^~~~~~
make[3]: *** [F-Prime/Fw/Types/CMakeFiles/Fw_Types.dir/PolyType.cpp.obj] Error 1
CMakeFiles/Makefile2:1673: recipe for target 'F-Prime/Fw/Types/CMakeFiles/Fw_Types.dir/all' failed
CMakeFiles/Makefile2:11460: recipe for target 'Primary/Top/CMakeFiles/Primary.dir/rule' failed
make[2]: *** [F-Prime/Fw/Types/CMakeFiles/Fw_Types.dir/all] Error 2
Makefile:3251: recipe for target 'Primary' failed
make[1]: *** [Primary/Top/CMakeFiles/Primary.dir/rule] Error 2
make: *** [Primary] Error 2
[ERROR] CMake erred with return code 2



We determined the source of the error by replacing the macros with what should be their evaluated values (i.e. replace PRIu8 with 'hhu' and we receive the same compile time error. Replace PRIu8 with "hhu" and the compile time error disappears)

**Comment by timcanham** [comment]
It is almost acting like it doesn't see the macros for `PRI*` and is treating it as an identifier. Then it would expect the `sprintf()` arguments to obey a function invocation rather than the macro substitution.

**Reply by teddylambert** [reply]
Yep, it looks like seeing the macros is the issue. Traced this a little further, and looks like it has to do with the `inttypes.h` being included by Arduino. (For reference, I am on Arduino 1.8.10)

The Arduino version of `inttypes.h` is definitely the one being included, because removing/renaming this file throws an error on the `#include inttypes.h` line saying that it isn't found.

However, within this version of `intypes.h`, the `PRI*` macros we want are wrapped within 2 `#define` statements:
1. `#if !defined(__cplusplus) || defined(__STDC_LIMIT_MACROS)` -- This wraps all of the macros we want, so with it none of the macros get defined
2. `#ifdef __avr_libc_does_not_implement_long_long_in_printf_or_scanf` -- This specifically wraps around the 64 bit macros, so even if the first `#ifdef` is removed we still get errors on the `PRI*64` macros specifically.

With both of these `#ifdef` statements removed, the original error disappears.

To me, it seems like having to edit this file included by the Arduino application itself is not the right way to go about things, and to instead focus on the conditions causing the `#ifdef` statements. Not sure about the best way to go about resolving this, but just wanted to provide an update on our debugging progress in case anyone has seen something like this before and has suggestions.

**Reply by Joshua-Anderson** [reply]
> #if !defined(__cplusplus) || defined(__STDC_LIMIT_MACROS) -- This wraps all of the macros we want, so with it none of the macros get defined

It looks like `__STDC_LIMIT_MACROS` isn't being defined when it should be. Is there another compilation flag that causes the definition of this flag?

> #ifdef __avr_libc_does_not_implement_long_long_in_printf_or_scanf -- This specifically wraps around the 64 bit macros, so even if the first #ifdef is removed we still get errors on the PRI*64 macros specifically.

That's ok... you can disable usage of 64 bit variables in F' by defining FW_HAS_F64=0

**Comment by SterlingPeet** [comment]
It is worth noting that this issue has disappeared when switching the F Prime version from ``devel`` to ``v2.0.1``

**Reply by Joshua-Anderson** [reply]
The change introducing the use of `PRI*` macros for formatting polytype objects occurred after the 2.0 release.... this will be coming in the next F' release. It would be great to get to the bottom of what's causing this problem on Arduino beforehan!
